packages/react-reconciler/src/ReactFiberHydrationContext.new.js

```diff
@@ -675,10 +692,22 @@ function getIsHydrating(): boolean {
 
 export function queueHydrationError(error: mixed): void {
   if (hydrationErrors === null) {
+    // We always queue the first hydration error
     hydrationErrors = [error];
   } else {
-    hydrationErrors.push(error);
+    // didThrowNotYetQueuedHydrationMismatchError will be true if we just threw
+    // a hydration mismatch error. In those cases we do not want to queue the
+    // error because there is a more useful error related to hydration that
+    // was already queued.
+    if (!didThrowNotYetQueuedHydrationMismatchError) {
```

Comment:
I’m not sure about this. Why do we need this? There’s something about special casing mismatches that doesn’t sit right with me. Partly because we’d have to remember which ones to tag and this is a very specific set of rules that we have to remember (eg only throws, not console errors, if we change something to no longer be a runtime error we have to also change this flag) and it should only be used if the mismatch could be due to not advancing the sibling.

Reply:
My general feeling is that if we warn we should error and if we don’t warn we shouldn’t error. It’s unfortunate that the parity here is coincidental rather than structural but if the logic makes sense to suppress mismatch warnings after the first one then I think it makes sense to extend to errors.

If there is any kind of error during hydration it could be reasonable to halt work and switch to client rendering mode immediately in which case there would be no more mismatch errors by definition. The choice to warm up components before client rendering is an implementation detail and thus suppressing the errors from them makes sense to me. If there is a problem the only rational debugging process it seems is to start with the first error (that has a preceding warning) and once fixed either there will be no more errors or the next problem will be uncovered. If we show the other mismatch errors they will generally be of no use to this kind of process.


One caveat is maybe including the first warning and error after successful hydration. So if hydration can somehow recover and then there is a new mismatch error we could included that. I didn’t go that route bc it felt incremental in value and the larger refactor of hydration would need to reimagine these errors anyway

Reply:
> If there is any kind of error during hydration it could be reasonable to halt work and switch to client rendering mode immediately in which case there would be no more mismatch errors by definition. The choice to warm up components before client rendering is an implementation detail and thus suppressing the errors from them makes sense to me. If there is a problem the only rational debugging process it seems is to start with the first error (that has a preceding warning) and once fixed either there will be no more errors or the next problem will be uncovered. If we show the other mismatch errors they will generally be of no use to this kind of process.

This whole argument makes sense for all errors, but there's nothing in here that says it should only apply to this subset of errors. IMO, it's not worth the complexity of special casing this subset when there are a bunch of other classes of errors that are not.

Reply:
Hmm. If there were two independent user errors within a suspense boundary wouldn't you expect to report both of them? the first error isn't necessarily going to imply anything about the second error? The causal link between later errors being ignorable after an initial one is that these later errors are caused by the earlier one but that IMO makes sense only for hydration mismatch errors, not errors in general. Why do you feel like all errors should follow a similar logic?

Is it just that we could choose not to continue rendering and in that case you wouldn't discover the second error until you fixed the first?

Reply:
A mismatch earlier doesn't mean that a mismatch later is also a false. It could be two mismatches.

Similarly, a regular error later doesn't mean it's not the same cause as the first error. So you can get spammed by multiple errors from a single cause anyway.

Either just because it's a shared errors but also due to purely ordering like using a technique like the old `useID` where you use an incrementing counter during render. Any render-impurity really can cause it.

Once imperative code kicks in you'll likely have a bunch of errors following a first one anyway. So the rule is always to look for the first one as the root cause.

React's implementation details is just one of many implementation details that can cause this.

The important rule is that the order is preserved. I.e. first displayed should always be the one that happened first.

If noise is an argument then we should drop all but one error.

Reply:
Mainly I care because reducing special case means reducing complexity. Especially this one I fear we'll screw up by making DEV/PROD differ that can be much worse.

Reply:
I'm not convinced that all errors are actually equivalent (enough) but I see the argument for simplicity. I'll update to keep only the first error. the complication is that we also use hydrationErrors to queue the descriptive error that reports that the boundary was rendered on client as a fallback. It uses the current hydrationErrors path and I think we would want something like that still so it won't be as simple as just storing the first queued error I think